### ISSUE TITLE
Hyperparameter optimization 1: Using Grid Search to Optimize Hyperparameters

### ISSUE THREAD

**Issue by collins-a** [section-engineering-education/engineering-education#1639]
### Proposed title of article

Machine Learning Hyperparameter Optimization 1: Using Grid Search to Optimize Hyperparameters

### Introduction paragraph(s):

Unlike parameters, hyperparameters are set before training a machine learning model. These hyperparameters need to be optimized in order to adapt a model to a dataset. However, the optimal hyperparameter settings on one dataset are unlikely to be optimal on another. This makes the task of hyperparameter optimization that much trickier. This is the first of a series of articles where we dive deep into the key techniques to optimize hyperparameters.

Grid search refers to a technique used to identify the optimal hyperparameters for a model. Unlike parameters, finding hyperparameters in training data is unattainable. As such, to find the right hyperparameters, we create a model for each combination of hyperparameters. Grid search is thus considered a very traditional hyperparameter optimization method since we are basically “brute-forcing” all possible combinations. The models are then evaluated through cross-validation. The model boasting the best accuracy is naturally considered to be the best. 

The example given below is a basic implementation of grid search. We simply first specify the hyperparameters we seek to examine. Then we provide a set of values to test. After this, grid search will attempt all possible hyperparameter combinations with the aid of cross-validation. Let’s break down this process into the steps below.

### Key takeaways: 

The reader should understand

1. The theory behind Grid Search
2. The role of cross-validation in hyperparameter optimization 
3. The implementation of Grid Search through a simple step by step example

### References: 

https://scikit-learn.org/stable/modules/grid_search.html#grid-search

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

https://scikit-learn.org/stable/glossary.html

https://machinelearningmastery.com/hyperparameter-optimization-with-random-search-and-grid-search/

https://blog.usejournal.com/a-comparison-of-grid-search-and-randomized-search-using-scikit-learn-29823179bc85

https://medium.com/datadriveninvestor/an-introduction-to-grid-search-ff57adcc0998

**Comment by ninjaginja** [comment]
Good topic @collins-a. However, please structure this as a standalone article (i.e. "Using Grid Search to Optimize Hyperparameters") vs. the first in a series of articles. We can't guarantee that all suggested future articles will be published. Thank you.
